README.rst
FLOSCRAPER
##########

Some basic webscraper I use in many projects.

.. image:: https://img.shields.io/pypi/v/floscraper.svg
    :target: https://pypi.python.org/pypi/floscraper

.. image:: https://img.shields.io/pypi/l/floscraper.svg
    :target: https://pypi.python.org/pypi/floscraper

.. image:: https://img.shields.io/pypi/dm/floscraper.svg
    :target: https://pypi.python.org/pypi/floscraper


webscraper
==========
Module to ease web efforts

**Supports**

* Cached web requests (Wrapper around requests)
* Bultin parsing/scraping (Wrapper around beautifulsoup)


**Constructor parameters**

* url: Default url, used if nothing else specified
* scheme: Default scheme for scrapping
* timeout
* cache_directory: Where to save cache files
* cache_time: How long is a cached resource vaild - in seconds (default: 7 minutes)
* cache_use_advanced
* auth_method: Authentication method (default: HTTPBasicAuth)
* auth_username: Authentication username. If set, enables authentication
* auth_password: Authentication password
* handle_redirect: Allow redirects (default: True)
* user_agent: User agent to use
* default_user_agents_browser: Browser to set in user agent (from ``default_user_agents`` dict)
* default_user_agents_os: Operating system to set in user agent (from ``default_user_agents`` dict)
* user_agents_browser: Browser to set in user agent (Overwrites default_user_agents_browser)
* user_agents_os: Operating system to set in user agent (Overwrites default_user_agents_os)
* html2text: HTML2text settings
* html_parser: What html parser to use (default: html.parser - built in)


**Example**

.. code-block:: python

    # Setup WebScraper with caching
    web = WebScraper({
        'cache_directory': "cache",
        'cache_time': 5*60
    })
    
    # First call to git -> hit internet
    web.get("https://github.com/")
    
    # Second call to git (within 5 minutes of first) -> hit cache
    web.get("https://github.com/")

Whitch results in the following output:

::

    2016-01-07 19:22:00 DEBUG   [WebScraper._getCached] From inet https://github.com
    2016-01-07 19:22:00 INFO    [requests.packages.urllib3.connectionpool] Starting new HTTPS connection (1): github.com
    2016-01-07 19:22:01 DEBUG   [requests.packages.urllib3.connectionpool] "GET / HTTP/1.1" 200 None
    2016-01-07 19:22:01 DEBUG   [WebScraper._getCached] From cache https://github.com


HISTORY.rst
.. :changelog:

History
=======

0.3.1 (2019-08-04)
---------------------

* Fix __init__
* Hashlib needs byte input
* Cache hit/miss info


0.3.0 (2019-08-03)
---------------------

* Upgrade flotils
* Remove tzinfo (default is utc)


0.2.3 (2018-07-02)
---------------------

* Upgrade flotils


0.2.2 (2017-12-07)
---------------------

* Fix cache duration bug


0.2.1 (2017-11-03)
---------------------

* Add raw response to unchached response


0.2.0 (2017-10-12)
---------------------

* Rework api names
* Redesign caching


0.1.15a0 (2016-03-08)
---------------------

* First release on PyPI.


floscraper/cache.py
# -*- coding: UTF-8 -*-
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

__author__ = "d01"
__email__ = "jungflor@gmail.com"
__copyright__ = "Copyright (C) 2017-19, Florian JUNG"
__license__ = "MIT"
__version__ = "0.2.1"
__date__ = "2019-08-04"
# Created: 2017-10-07 15:19

import datetime
import os
import abc
import threading
import logging
import hashlib
from io import open

from flotils import Loadable

try:
    import portalocker as porta
except ImportError:
    # Not using portalocker
    porta = None

from .models import CacheInfo


_cache = {}
""" temp cache """
_cache_lock = threading.RLock()

if porta is None:
    logging.warning("Not using portalocker")


def now_utc():
    """
    Get current time as utc with tzinfo

    :return: Current time
    :rtype: datetime.datetime
    """
    return datetime.datetime.utcnow()


class Cache(Loadable):
    """ Cache element """
    __metaclass__ = abc.ABCMeta

    def __init__(self, settings=None):
        if settings is None:
            settings = {}
        super(Cache, self).__init__(settings)
        self._duration = datetime.timedelta()
        self.duration = settings.get('duration', 7 * 60)
        self.use_advanced = settings.get('use_advanced', True)

    @property
    def duration(self):
        return self._duration

    @duration.setter
    def duration(self, value):
        if not isinstance(value, datetime.timedelta):
            value = datetime.timedelta(seconds=value)
        self._duration = value

    def prepare_headers(self, headers, cache_info=None):
        """
        Prepare headers object for request (add cache information

        :param headers: Headers object
        :type headers: dict
        :param cache_info: Cache information to add
        :type cache_info: floscraper.models.CacheInfo
        :return: Prepared headers
        :rtype: dict
        """
        if self.use_advanced and cache_info:
            hkeys = headers.keys()
            if cache_info.access_time and "If-Modified-Since" not in hkeys:
                headers['If-Modified-Since'] = cache_info.access_time.strftime(
                    "%a, %d %b %Y %H:%M:%S GMT"
                )
            if cache_info.etag and "If-None-Match" not in hkeys:
                headers['If-None-Match'] = cache_info.etag
        return headers

    @abc.abstractmethod
    def _cache_meta_get(self, key):
        raise NotImplementedError()

    @abc.abstractmethod
    def _cache_get(self, key):
        raise NotImplementedError()

    @abc.abstractmethod
    def _cache_meta_set(self, key, val):
        raise NotImplementedError()

    @abc.abstractmethod
    def _cache_set(self, key, val):
        raise NotImplementedError()

    def get(self, url, ignore_access_time=False):
        """
        Try to retrieve url from cache if available

        :param url: Url to retrieve
        :type url: str | unicode
        :param ignore_access_time: Should ignore the access time
        :type ignore_access_time: bool
        :return: (data, CacheInfo)
            None, None -> not found in cache
            None, CacheInfo -> found, but is expired
            data, CacheInfo -> found in cache
        :rtype: (None | str | unicode, None | floscraper.models.CacheInfo)
        """
        key = hashlib.md5(url.encode("utf-8")).hexdigest()
        accessed = self._cache_meta_get(key)

        if not accessed:
            # Not previously cached
            self.debug("From inet {}".format(url))
            return None, None

        if isinstance(accessed, dict):
            cached = CacheInfo.from_dict(accessed)
        else:
            cached = CacheInfo(accessed)
        now = now_utc()
        if now - cached.access_time > self.duration and not ignore_access_time:
            # Cached expired -> remove
            self.debug("From inet (expired) {}".format(url))
            return None, cached

        try:
            res = self._cache_get(key)
        except:
            self.debug("From inet (failure) {}".format(url))
            self.exception("Failed to read cache")
            return None, None
        self.debug("From cache {}".format(url))
        return res, cached

    def update(self, url, cache_info=None):
        """
        Update cache information for url

        :param url: Update for this url
        :type url: str | unicode
        :param cache_info: Cache info
        :type cache_info: floscraper.models.CacheInfo
        :rtype: None
        """
        key = hashlib.md5(url.encode("utf-8")).hexdigest()
        access_time = None
        if not cache_info:
            cache_info = CacheInfo()
        if not access_time:
            cache_info.access_time = now_utc()
        self._cache_meta_set(key, cache_info.to_dict())

    def put(self, url, html, cache_info=None):
        """
        Put response into cache

        :param url: Url to cache
        :type url: str | unicode
        :param html: HTML content of url
        :type html: str | unicode
        :param cache_info: Cache Info (default: None)
        :type cache_info: floscraper.models.CacheInfo
        :rtype: None
        """
        key = hashlib.md5(url.encode("utf-8")).hexdigest()

        try:
            self._cache_set(key, html)
        except:
            self.exception("Failed to write cache")
            return
        self.update(url, cache_info)


class NullCache(Cache):
    """ Non caching cache """

    def _cache_meta_get(self, key):
        pass

    def _cache_get(self, key):
        pass

    def _cache_set(self, key, val):
        pass

    def _cache_meta_set(self, key, val):
        pass

    def get(self, url, ignore_access_time=False):
        return None, None

    def update(self, url, cache_info=None):
        pass

    def put(self, url, html, cache_info=None):
        pass


class FileCache(Cache):

    def __init__(self, settings=None):
        if settings is None:
            settings = {}
        super(FileCache, self).__init__(settings)
        self._dir = settings['directory']
        self._index_path = settings.get(
            'index', os.path.join(self._dir, "cache_index.tmp")
        )
        self._index = None
        with _cache_lock:
            self._index = _cache

    def _init_index(self):
        if not self._index:
            if os.path.exists(self._index_path):
                try:
                    with open(self._index_path, "r", encoding="utf-8") as f:
                        if porta:
                            porta.lock(f, porta.LOCK_EX)
                        self._index = self._load_json_file(f)
                except IOError as e:
                    if str(e) != "Decoding json failed":
                        self.exception("Failed to load cache file")
                except:
                    self.exception("Failed to load cache file")
            else:
                self.warning(
                    "Cache index does not exist ({})".format(self._index_path)
                )

        if not self._index:
            self._index = {'version': "1.0"}

    def _cache_meta_get(self, key):
        with _cache_lock:
            _cache.update(self._index)
            return _cache.get(key, None)

    def _cache_get(self, key):
        tmp_path = os.path.join(self._dir, key + ".tmp")
        with open(tmp_path, "r", encoding="utf-8") as f:
            if porta:
                porta.lock(f, porta.LOCK_EX)
            return f.read()

    def _cache_meta_set(self, key, val):
        self._index[key] = val
        with _cache_lock:
            _cache[key] = self._index[key]

    def _cache_set(self, key, val):
        tmp_path = os.path.join(self._dir, key + ".tmp")
        with open(tmp_path, "w", encoding="utf-8") as f:
            if porta:
                porta.lock(f, porta.LOCK_EX)
            f.write(val)
            # Try flushing to make data available sooner
            # (try to get rid of erroneous reads)
            f.flush()

    def get(self, url, ignore_access_time=False):
        if not self._dir:
            self.debug("From inet {}".format(url))
            return None, None, None

        self._init_index()

        return super(FileCache, self).get(url, ignore_access_time)

    def put(self, url, html, cache_info=None):
        if not self._dir:
            return
        super(FileCache, self).put(url, html, cache_info)

        # Better in a close statement?
        try:
            with open(
                    self._index_path, "w", encoding="utf-8"
            ) as f, _cache_lock:
                if porta:
                    porta.lock(f, porta.LOCK_EX)
                self._save_json_file(f, _cache)
                # try flushing to make data available sooner
                # (try to get rid of erroneous reads)
                f.flush()
        except:
            self.exception("Failed to save cache")



floscraper/models.py
# -*- coding: UTF-8 -*-
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

__author__ = "the01"
__email__ = "jungflor@gmail.com"
__copyright__ = "Copyright (C) 2017-19, Florian JUNG"
__license__ = "MIT"
__version__ = "0.1.2"
__date__ = "2018-08-04"
# Created: 2017-10-06 23:35

import flotils


class CacheInfo(flotils.FromToDictBase, flotils.PrintableBase):
    """ Cache information """

    def __init__(self, access_time=None, etag=None):
        super(CacheInfo, self).__init__()
        self.etag = etag
        """ :type : None | str | unicode """
        self.access_time = access_time
        """ :type : None | datetime.datetime """
        self.hit = None
        """ Cache hit
            :type : None | bool """


class Response(flotils.FromToDictBase, flotils.PrintableBase):
    """ Scrapper response object """

    def __init__(self, html=None, cache_info=None, scraped=None, raw=None):
        super(Response, self).__init__()
        self.cache_info = cache_info
        """ :type : None | CacheInfo """
        self.raw = raw
        """ Raw, undecoded reponse
            :type : None | unicode """
        self.html = html
        """ Html reponse
            :type : None | unicode """
        self.scraped = scraped
        """ Scrapped content
            :type : None | list | dict """

    def __str__(self):
        return "({}), {}, {}, {}".format(
            self.cache_info, self.html, self.scraped, self.raw
        )

    def to_dict(self):
        """
        Response as dict

        :return: response
        :rtype: dict
        """
        res = super(Response, self).to_dict()

        if self.cache_info:
            res['cache_info'] = self.cache_info.to_dict()
        return res

    @staticmethod
    def from_dict(d):
        """
        Response from dict

        :param d: Dict to load
        :type d: dict
        :return: response
        :rtype: Response
        """
        if d is None:
            return None
        return Response(
            d.get('html'),
            CacheInfo.from_dict(d.get('cache_info')),
            d.get('scraped'),
            d.get('raw')
        )


floscraper/webscraper.py
# -*- coding: UTF-8 -*-
"""
Module for loading/scraping data from the web
"""
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

__author__ = "the01"
__email__ = "jungflor@gmail.com"
__copyright__ = "Copyright (C) 2014-19, Florian JUNG"
__license__ = "MIT"
__version__ = "0.1.11"
__date__ = "2019-08-04"
# Created: 2014-04-02 11:23

import re
import socket

from bs4 import BeautifulSoup
import html2text
import requests
from requests import HTTPError
from requests.exceptions import SSLError, Timeout, ConnectionError

from flotils.loadable import Loadable

from .default_user_agents import default_user_agents
from .models import Response, CacheInfo
from .cache import FileCache, NullCache


class WEBParameterException(Exception):
    """ Parameter Exception """
    pass


class WEBFileException(IOError):
    """ File Exception """
    pass


class WEBConnectException(Exception):
    """ Error with web connection """
    pass


class WebScraper(Loadable):
    """ Class for cached, session get/post/.. with optional scraping """

    def __init__(self, settings=None):
        """
        Initialize object

        :param settings: Settings for instance (default: None)
        :type settings: dict | None
        :rtype: None
        """
        if settings is None:
            settings = {}
        super(WebScraper, self).__init__(settings)

        self.url = settings.get('url', None)
        self.scheme = settings.get('scheme', None)
        self.timeout = settings.get('timeout', None)

        cache_sett = settings.get('cache')
        self.cache = NullCache()
        """ :type : None | floscraper.cache.Cache """
        if cache_sett:
            self.cache = FileCache(cache_sett)

        self._auth_method = settings.get('auth_method', None)
        self._auth_username = settings.get('auth_username', None)
        self._auth_password = settings.get('auth_password', None)

        self.session = None
        """ Object to do http actions
            :type _br: requests.Session """
        self._handle_redirect = settings.get('handle_redirect', True)

        agent_browser = default_user_agents['browser']['keith']
        agent_os = default_user_agents['os']['linux_i686']
        agent = settings.get('user_agent', None)

        if "default_user_agents_browser" in settings:
            brow = settings['default_user_agents_browser']
            agent_browser = default_user_agents['browser'].get(brow, None)
        if "default_user_agents_os" in settings:
            os = settings['default_user_agents_os']
            agent_os = default_user_agents['os'].get(os, None)
        if "user_agent_browser" in settings:
            agent_browser = settings['user_agent_browser']
        if "user_agent_os" in settings:
            agent_os = settings['user_agent_os']

        if not agent:
            if agent_browser and agent_os:
                agent = agent_browser.format(agent_os)
        self.user_agent = agent
        self._text_maker = None
        """ Object to translate html to markdown (html2text)
            :type : None | html2text.HTML2Text """
        if settings.get('html2text'):
            self._set_html2text(settings['html2text'])
        self.html_parser = settings.get('html_parser', "html.parser")
        """ What html parser to use (default: html.parser - built in)
            :type : str | unicode """

    def _browser_init(self):
        """
        Init the browsing instance if not setup

        :rtype: None
        """
        if self.session:
            return

        self.session = requests.Session()
        headers = {}

        if self.user_agent:
            headers['User-agent'] = self.user_agent
        self.session.headers.update(headers)

        if self._auth_method in [None, "", "HTTPBasicAuth"]:
            if self._auth_username is not None:
                self.session.auth = (self._auth_username, self._auth_password)

    def _set_html2text(self, settings):
        """
        Load settings for html2text (https://github.com/Alir3z4/html2text)

        Warning: does not check options/values

        :param settings: Settings for the object
            (see:https://github.com/Alir3z4/html2text/blob/master/docs/usage.md)
        :type settings: dict
        :rtype: None
        """
        self._text_maker = html2text.HTML2Text()
        for param in settings:
            if not hasattr(self._text_maker, param):
                raise WEBParameterException(
                    "Setting html2text failed - unknown parameter {}".format(
                        param
                    )
                )
            setattr(self._text_maker, param, settings[param])

    def load_scrap(self, path):
        """
        Load scraper settings from file

        :param path: Path to file
        :type path: str
        :rtype: None
        :raises WEBFileException: Failed to load settings
        :raises WEBParameterException: Missing parameters in file
        """
        try:
            conf = self.load_settings(path)
        except:
            # Should only be IOError
            self.exception("Failed to load file")
            raise WEBFileException("Failed to load from {}".format(path))

        if "scheme" not in conf:
            raise WEBParameterException("Missing scheme definition")
        if "url" not in conf:
            raise WEBParameterException("Missing url definition")
        version = conf.get('version', None)
        if version != "1.0":
            raise WEBParameterException(
                "Unsupported version {}".format(version)
            )
        self.scheme = conf['scheme']
        self.url = conf['url']
        self.timeout = conf.get('timeout', self.timeout)
        if conf.get('html2text'):
            self._set_html2text(conf['html2text'])

    def request(
            self, method, url, timeout=None,
            headers=None, data=None, params=None
    ):
        """
        Make a request using the requests library

        :param method: Which http method to use (GET/POST)
        :type method: str | unicode
        :param url: Url to make request to
        :type url: str | unicode
        :param timeout: Timeout for request (default: None)
            None -> infinite timeout
        :type timeout: None | int | float
        :param headers: Headers to be passed along (default: None)
        :type headers: None | dict
        :param data: Data to be passed along (e.g. body in POST)
            (default: None)
        :type data: None | dict
        :param params: Parameters to be passed along (e.g. with url in GET)
            (default: None)
        :type params: None | dict
        :return: Response to the request
        :rtype: requests.Response
        :raises WEBConnectException: Loading failed
        """
        if headers is None:
            headers = {}
        if not self.session:
            self._browser_init()

        try:
            response = self.session.request(
                method,
                url,
                timeout=timeout,
                allow_redirects=self._handle_redirect,
                headers=headers,
                data=data,
                params=params
            )
        except SSLError as e:
            raise WEBConnectException(e)
        except HTTPError:
            raise WEBConnectException("Unable to load {}".format(url))
        except (Timeout, socket.timeout):
            raise WEBConnectException("Timeout loading {}".format(url))
        except ConnectionError:
            raise WEBConnectException("Failed to load {}".format(url))
        except Exception:
            self.exception("Failed to load {}".format(url))
            raise WEBConnectException(
                "Unknown failure loading {}".format(url)
            )
        return response

    def _get(self, url, **kwargs):
        """
        Make GET request

        :param url: Url to make the request to
        :type url: str | unicode
        :param kwargs: See _requests for parameters
        :type kwargs:
        :return: Response
        :rtype: requests.Response
        """
        return self.request("GET", url, **kwargs)

    def _post(self, url, **kwargs):
        """
        Make POST request

        :param url: Url to make the request to
        :type url: str | unicode
        :param kwargs: See _requests for parameters
        :type kwargs: dict
        :return: Response
        :rtype: requests.Response
        """
        return self.request("POST", url, **kwargs)

    def get(self, url, timeout=None, headers=None, params=None, cache_ext=None):
        """
        Make get request to url (might use cache)

        :param url: Url to make request to
        :type url: str | unicode
        :param timeout: Timeout for request (default: None)
            None -> infinite timeout
        :type timeout: None | int | float
        :param headers: Headers to be passed along (default: None)
        :type headers: None | dict
        :param params: Parameters to be passed along with url (default: None)
        :type params: None | dict
        :param cache_ext: External cache info
        :type cache_ext: floscraper.models.CacheInfo
        :return: Response
        :rtype: floscraper.models.Response | None
        :raises WEBConnectException: Loading failed
        """
        if headers is None:
            headers = {}
        cached = cache_info = None
        # TODO: add params to caching key
        if self.cache:
            cached, cache_info = self.cache.get(url)

        if cache_ext:
            # Check local cache
            if not cache_info:
                cache_info = cache_ext
            if cache_info.access_time and cache_ext.access_time:
                if cache_info.access_time < cache_ext.access_time:
                    cached = None
                    cache_info = cache_ext
            if cache_info.etag != cache_ext.etag:
                cached = None
                cache_info = cache_ext

        if cached:
            # Using cached
            if cache_info:
                cache_info.hit = True
            return Response(cached, cache_info)
        if cache_info:
            cache_info.hit = None
        # Not using cached

        if not self.session:
            self._browser_init()

        if self.cache:
            headers = self.cache.prepare_headers(headers, cache_info)

        response = self._get(
            url,
            timeout=timeout,
            headers=headers,
            params=params
        )
        if "etag" in response.headers:
            if not cache_info:
                cache_info = CacheInfo()
            cache_info.etag = response.headers.get('etag')

        res = Response(cache_info=cache_info)
        if cache_info:
            cache_info.hit = False

        if response.history:
            # list of responses in redirects
            code = 0

            for resp in response.history + [None]:
                if resp is None:
                    resp = response
                if code in [
                    requests.codes.FOUND,
                    requests.codes.SEE_OTHER,
                    requests.codes.TEMPORARY_REDIRECT
                ]:
                    self.info("Temporary redirect to {}".format(resp.url))
                elif code in [
                    requests.codes.MOVED_PERMANENTLY,
                    requests.codes.PERMANENT_REDIRECT
                ]:
                    self.warning("Moved to {}".format(resp.url))
                elif code != 0:
                    self.error("Code {} to {}".format(code, resp.url))
                code = resp.status_code

        if response.status_code == requests.codes.NOT_MODIFIED:
            self.info("Not modified {}".format(url))
            res.html, _ = self.cache.get(url, ignore_access_time=True)
            if cache_info:
                cache_info.hit = True
            if self.cache:
                self.cache.update(url, cache_info)
            return res

        try:
            response.raise_for_status()
        except HTTPError as e:
            raise WEBConnectException("{} - {}".format(e, url))

        try:
            raw = response.content
            html = response.text
            if raw is None:
                self.warning("Response returned None")
                raise Exception()
            if html is None:
                self.warning("Response parsed returned None")
                raise Exception()
        except Exception:
            raise WEBConnectException("Unable to load {}".format(url))

        # TODO: cache raw / whole response object
        if self.cache:
            self.cache.put(url, html, cache_info)
        if url != response.url:
            if not response.history:
                self.warning(
                    "Response url different despite no redirects "
                    "{} - {}".format(url, response.url)
                )
            if self.cache:
                self.cache.put(response.url, html, cache_info)
        res.html = html
        res.raw = raw
        return res

    def _get_tag_match(self, ele, tree):
        """
        Match tag

        :param ele:
        :type ele:
        :param tree:
        :type tree: None, list
        :return:
        :rtype: None | list
        """
        if tree in [None, []]:
            return [ele]

        res = []
        t = tree[0]
        branch = tree[1:]
        attributes = {}

        for attr in t:
            if isinstance(t[attr], dict):
                if t[attr].get("type", None) == "reg":
                    t[attr] = re.compile(t[attr]['reg'])

        attributes.update(t)

        if "name" in attributes:
            del attributes['name']
        if "text" in attributes:
            del attributes['text']
        if "recursive" in attributes:
            del attributes['recursive']
        if "[]" in attributes:
            del attributes['[]']

        possibles = ele.find_all(
            t.get('name', None),
            text=t.get('text', None),
            attrs=attributes,
            recursive=t.get('recursive', True)
        )

        if not possibles:
            return None
        else:
            pass

        if "[]" in t:
            try:
                possibles = eval("possibles[{}]".format(t["[]"]))
            except:
                # no possibles
                return None

        if not isinstance(possibles, list):
            possibles = [possibles]

        for a in possibles:
            match = self._get_tag_match(a, branch)

            if match:
                res.extend(match)

        if not res:
            return None
        else:
            return res

    def _parse_value(self, eles, value_scheme):
        """

        :param eles:
        :type eles: list
        :param value_scheme:
        :type value_scheme: dict
        :return:
        :rtype: list
        """
        val = []
        val_type = value_scheme.get('type', None)
        reg = value_scheme.get('reg', None)
        strip = value_scheme.get('strip', False)

        for match in eles:
            if val_type == "text":
                val.append("{}".format(match.getText()))
            elif val_type == "content":
                val.append("{}".format(match.getText()))
            elif val_type == "attribute" and value_scheme.get("attribute", None):
                attr = value_scheme['attribute']

                if attr in match.attrs:
                    res = match[attr]

                    # multi-valued attributes will be joined in one string
                    if isinstance(res, list):
                        res = " ".join(res)
                    val.append(res)
            elif val_type == "html":
                val.append("{}".format(match))
            else:
                # == html2text
                if self._text_maker is None:
                    val.append(html2text.html2text("{}".format(match)))
                else:
                    val.append(self._text_maker.handle("{}".format(match)))
            if strip:
                # TODO: github date field
                val[-1] = val[-1].strip()

        if reg:
            res = []

            for a in val:
                if isinstance(reg, dict):
                    if reg.get('type', None) == "reg":
                        reg = re.compile(reg['reg'])
                res.extend(reg.findall(a))
        else:
            res = val

        return res

    def _parse_scheme(self, ele, scheme):
        """

        :param ele:
        :type ele:
        :param scheme:
        :type scheme: dict[str | unicode, dict]
        :return:
        :rtype: dict
        """
        res = {}

        for aKey in scheme:
            entity = scheme[aKey]
            val = []
            matches = []
            res[aKey] = []

            if "tree" in entity:
                matches = self._get_tag_match(ele, entity['tree'])

                if not matches:
                    matches = []
                else:
                    pass
            else:
                # matches = [ele]
                pass

            if not matches and "value" == aKey:
                val.extend(self._parse_value([ele], entity))
                res[aKey].extend(val)

            if "value" == aKey:
                for a in matches:
                    # TODO: sure it's ele and not a?
                    val.extend(self._parse_value([ele], entity))
            if "children" in entity:
                for a in matches:
                    obj = {}
                    if "value" == aKey:
                        obj['value'] = val
                    child = self._parse_scheme(a, entity['children'])

                    if child:
                        obj.update(child)
                    if obj:
                        res[aKey].append(obj)
        return res

    def scrap(self,
              url=None, scheme=None, timeout=None,
              html_parser=None, cache_ext=None
    ):
        """
        Scrap a url and parse the content according to scheme

        :param url: Url to parse (default: self._url)
        :type url: str
        :param scheme: Scheme to apply to html (default: self._scheme)
        :type scheme: dict
        :param timeout: Timeout for http operation (default: self._timout)
        :type timeout: float
        :param html_parser: What html parser to use (default: self._html_parser)
        :type html_parser: str | unicode
        :param cache_ext: External cache info
        :type cache_ext: floscraper.models.CacheInfo
        :return: Response data from url and parsed info
        :rtype: floscraper.models.Response
        :raises WEBConnectException: HTTP get failed
        :raises WEBParameterException: Missing scheme or url
        """
        if not url:
            url = self.url
        if not scheme:
            scheme = self.scheme
        if not timeout:
            timeout = self.timeout
        if not html_parser:
            html_parser = self.html_parser
        if not scheme:
            raise WEBParameterException("Missing scheme definition")
        if not url:
            raise WEBParameterException("Missing url definition")
        resp = self.get(url, timeout, cache_ext=cache_ext)
        soup = BeautifulSoup(resp.html, html_parser)
        resp.scraped = self._parse_scheme(soup, scheme)
        return resp

    def _shrink_list(self, shrink):
        """
        Shrink list down to essentials

        :param shrink: List to shrink
        :type shrink: list
        :return: Shrunk list
        :rtype: list
        """
        res = []

        if len(shrink) == 1:
            return self.shrink(shrink[0])
        else:
            for a in shrink:
                temp = self.shrink(a)

                if temp:
                    res.append(temp)
        return res

    def _shrink_dict(self, shrink):
        """
        Shrink dict down to essentials

        :param shrink: Dict to shrink
        :type shrink: dict
        :return: Shrunk dict
        :rtype: dict
        """
        res = {}

        if len(shrink.keys()) == 1 and "value" in shrink:
            return self.shrink(shrink['value'])
        else:
            for a in shrink:
                res[a] = self.shrink(shrink[a])

                if not res[a]:
                    del res[a]
        return res

    def shrink(self, shrink):
        """
        Remove unnecessary parts

        :param shrink: Object to shringk
        :type shrink: dict | list
        :return: Shrunk object
        :rtype: dict | list
        """
        if isinstance(shrink, list):
            return self._shrink_list(shrink)
        if isinstance(shrink, dict):
            return self._shrink_dict(shrink)
        return shrink
